### PR TITLE
Use SPDX license identifiers

### LINKED ISSUE #1679: [RFC] Use SPDX license identifiers
All alpaka headers use a license statement at the top. This adds a lot of noise and there is a more compact way to specify the license: SPDX license identifiers. More info here: https://spdx.dev/ids/

We basically replace the existing license preamble in header files by:
`// SPDX-License-Identifier: MPL-2.0`

Are there any concerns to such a change?

### PR DESCRIPTION
**Pull request by j-stephan**
Fixes #1679. Replaces all header license comments with the appropriate SPDX license identifiers. In most places, this is `MPL-2.0`. In the `examples` folder this is `ISC`.

### CODE REVIEW

File: example/LICENSE [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Copyright 2014 - 2023 Helmholtz-Zentrum Dresden - Rossendorf e.V. (HZDR)
```
Comment (j-stephan) [inline example/LICENSE]: Since alpaka started out as a HZDR research project I believe the formal copyright holder is HZDR. Please correct me if I'm wrong @alpaka-group/alpaka-maintainers.

File: cmake/addExecutable.cmake [hunk 2]
```diff
@@ -1,14 +1,9 @@
...
-# This file is part of alpaka.
-#
-# This Source Code Form is subject to the terms of the Mozilla Public
-# License, v. 2.0. If a copy of the MPL was not distributed with this
-# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# SPDX-License-Identifier: MPL-2.0
 #
 
-CMAKE_MINIMUM_REQUIRED(VERSION 3.18)
+CMAKE_MINIMUM_REQUIRED(VERSION 3.22)
```
Comment (bernhardmgruber) [inline cmake/addExecutable.cmake]: I think you mixed some unrelated changes into this PR.

File: example/LICENSE (hunk 1 above)
Comment (bernhardmgruber) [reply example/LICENSE]: I don't think this is right. This makes it look like the entire alpaka examples are copyrighted by HZDR. That is at least wrong for all LOCs I have written, which legally belong to CERN.

Why do you even need to put a LICENSE file here?

File: example/babelstream/CMakeLists.txt [hunk 3]
```diff
@@ -1,19 +1,6 @@
...
-# copyright notice and this permission notice appear in all copies.
-#
-# THE SOFTWARE IS PROVIDED “AS IS” AND ISC DISCLAIMS ALL WARRANTIES WITH
-# REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
-# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
-# SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
-# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
-# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
-# IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+# SPDX-License-Identifier: ISC
```
Comment (bernhardmgruber) [inline example/babelstream/CMakeLists.txt:3]: Have all the examples previously already had the ISC license? That would be news to me, I thought everything was MPL-2. In order words, are you relicensing all the examples?

File: example/LICENSE (hunk 1 above)
Comment (j-stephan) [reply example/LICENSE]: We should discuss the copyright issue in tomorrow's VC.

> Why do you even need to put a LICENSE file here?

We need to put the ISC license for the examples somewhere if we delete the header comments. Feel free to suggest a better location.

File: example/babelstream/CMakeLists.txt (hunk 3 above)
Comment (j-stephan) [reply example/babelstream/CMakeLists.txt:3]: The changes are shown right there ;-) As far as I can tell all examples except BabelStream have always been licensed under ISC. I didn't relicense anything (on purpose).

File: example/randomCells2D/src/randomCells2D.cpp [hunk 4]
```diff
@@ -1,10 +1,5 @@
 /* Copyright 2021 Jiri Vyskocil
- *
- * This file is part of alpaka.
- *
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * SPDX-License-Identifier: ISC
```
Comment (j-stephan) [inline example/randomCells2D/src/randomCells2D.cpp:2]: Here we have a case of accidental relicensing due to license mixing. The CMakeLists.txt is ISC-licensed but the source code is MPL. I attribute this to a copy-paste error but maybe @sliwowitz did this on purpose?

File: example/randomStrategies/src/randomStrategies.cpp [hunk 5]
```diff
@@ -1,10 +1,5 @@
 /* Copyright 2022 Jiří Vyskočil, René Widera, Jan Stephan
- *
- * This file is part of alpaka.
- *
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * SPDX-License-Identifier: ISC
```
Comment (j-stephan) [inline example/randomStrategies/src/randomStrategies.cpp:2]: And here, too.

File: example/LICENSE (hunk 1 above)
Comment (bernhardmgruber) [reply example/LICENSE]: > We need to put the ISC license for the examples somewhere if we delete the header comments. Feel free to suggest a better location.

Do we need to do that? The ISC license doesn't state that a copy of the license must be included with the source code. We could just have the ISC SPDX license identifier in the example files and be good I think.

File: example/randomCells2D/src/randomCells2D.cpp (hunk 4 above)
Comment (sliwowitz) [reply example/randomCells2D/src/randomCells2D.cpp:2]: Yes, it was a mistake. I did not realize we use different licenses for the examples.

Comment (sliwowitz) [reply example/randomCells2D/src/randomCells2D.cpp:2]: We can change that to ISC or whatever license we decide to use for the examples.

File: example/randomStrategies/src/randomStrategies.cpp (hunk 5 above)
Comment (sliwowitz) [reply example/randomStrategies/src/randomStrategies.cpp:2]: Same as the the other file.

**Comment by fwyzard** [comment]
Is there a good reason for the ISC license ?

According to [Wikipedia](https://en.wikipedia.org/wiki/ISC_license#Reception), ISC itself stopped using it, and moved to MPL 2.0 .

**Comment by j-stephan** [comment]
This was decided upon in 2018 when alpaka was relicensed from LGPLv3 to the current MPLv2/ISC mix. You can follow the chain of reasoning [here](https://github.com/alpaka-group/alpaka/issues/637) (I only discovered this yesterday while looking into the same question). Apparently ISC is "more modern" than BSD/MIT.

**Comment by fwyzard** [comment]
> Apparently ISC is "more modern" than BSD/MIT.

I don't contest that (though the differences seem minor at best).
But why not use MPL 2.0 everywhere, instead of a mix of MPL 2.0/ISC ?

Anyway, that's orthogonal to this PR... I am simply curious, and didn't know ISC was used for (some of ?) the examples.

**Comment by j-stephan** [comment]
Our predecessors argued that the examples should be copy&paste-able into any code (including closed source), hence ISC for the examples. But I'm no license expert, it is entirely possible that MPL would also allow for this.

File: include/alpaka/rand/TinyMT/tinymt32.h [hunk 6]
```diff
@@ -1,10 +1,5 @@
 /* Copyright 2020 Axel Huebl, Benjamin Worpitz, Mutsuo Saito, Bernhard Manfred Gruber
- *
- * This file is part of alpaka.
- *
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * SPDX-License-Identifier: MPL-2.0
```
Comment (bernhardmgruber) [inline include/alpaka/rand/TinyMT/tinymt32.h]:
```suggestion
 * SPDX-License-Identifier: BSD-3-Clause
```
I would generally replace the author list above by the the original authors of the file's main content.

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]

File: include/alpaka/rand/TinyMT/tinymt32.h (hunk 6 above)
Comment (j-stephan) [reply include/alpaka/rand/TinyMT/tinymt32.h]: Good catch. This was added automatically a few years back and is likely an oversight.

**Review by bernhardmgruber — APPROVED** [review]
